python311Packages.connectorx: init at 0.4.1

coastalwhite opened the pull request:
## Description of changes

This adds the [connectorx] python package. These are used by several of the data science and dataframe libraries in the python ecosystem. Specifically, these are packaged now with the hope of better packaging of the [polars] in the future.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc
[connectorx]: https://sfu-db.github.io/connector-x
[deltalake]: https://delta-io.github.io/delta-rs/
[polars]: https://github.com/pola-rs/polars

coastalwhite committed:
python311Packages.connectorx: init at 0.4.1

pluiedev reviewed (changes requested):
Please also rename the PR to fit contributing conventions: `pythonPackages.connectorx: init at 0.3.3; pythonPackages.deltalake: init at 0.17.4`

pluiedev commented on pkgs/development/python-modules/deltalake/default.nix:
```diff
@@ -0,0 +1,63 @@
+
+  buildInputs = [ openssl ];
+
+  pythonImportsCheck = [ pname ];
```
I was initially quite confused by this: is `pname` some package I didn't know of?
For clarity's sake, just specify the package name explicitly.

```suggestion
  pythonImportsCheck = [ "deltalake" ];
```

pluiedev commented on pkgs/development/python-modules/connectorx/default.nix:
```diff
@@ -0,0 +1,61 @@
+    openssl # needed for openssl-sys
+  ];
+
+  pythonImportsCheck = [ pname ];
```
Ditto

natsukium commented on pkgs/development/python-modules/connectorx/default.nix:
```diff
@@ -0,0 +1,61 @@
+  ];
+
+  pythonImportsCheck = [ "connectorx" ];
+
```
Could you please add `pytestCheckHook` to `nativeCheckInputs`?
https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pytestcheckhook-using-pytestcheckhook

natsukium reviewed (changes requested):
Please use `python311Packages` instead of `pythonPackages`.
`pythonPackages` means `python2` in nixpkgs.

harvidsen commented:
Note that deltalake already has a PR https://github.com/NixOS/nixpkgs/pull/320485.

coastalwhite committed:
maintainers: add coastalwhite

natsukium reviewed (changes requested):
Could you format it with `nixfmt-rfc-style` and look into the build error?

```
error: failed to load manifest for dependency `connectorx`
Caused by:
  failed to parse manifest at `/build/source/connectorx/Cargo.toml`
Caused by:
  dependency (serde) specified without providing a local path, Git repository, version, or workspace dependency to use
```

pluiedev commented on pkgs/development/python-modules/connectorx/default.nix:
```diff
@@ -0,0 +1,65 @@
+{ lib
+, buildPythonPackage
+, rustPlatform
+, pytestCheckHook
+, fetchFromGitHub
+, krb5-c
+, openssl
+, libkrb5
+,
```
Needs to be reformatted with `nixfmt-rfc-style``

pluiedev commented on pkgs/development/python-modules/connectorx/default.nix:
```diff
@@ -0,0 +1,65 @@
+  src = fetchFromGitHub {
+    owner = "sfu-db";
+    repo = "connector-x";
+    rev = "v${version}";
```
```suggestion
    tag = "v${version}";
```

pluiedev commented on pkgs/top-level/python-packages.nix:
```diff
@@ -2065,6 +2065,10 @@ self: super: with self; {
   connection-pool = callPackage ../development/python-modules/connection-pool { };
 
+  connectorx = callPackage ../development/python-modules/connectorx {
+    krb5-c = pkgs.krb5;
```
Is this necessary? (Why not just use `krb5` directly?)

coastalwhite replied:
This is because it will otherwise try to use `python3Packages.krb5`. This is following the comment from above <https://github.com/NixOS/nixpkgs/pull/312884#discussion_r1727169522>.

pluiedev commented on pkgs/development/python-modules/connectorx/default.nix:
```diff
@@ -0,0 +1,65 @@
+  pythonImportsCheck = [ "connectorx" ];
+
+  meta = {
+    description =
+      "Fastest library to load data from DB to DataFrames in Rust and Python";
```
```suggestion
    description = "Fastest library to load data from DB to DataFrames in Rust and Python";
```

coastalwhite commented:
Is there anything that still needs to be done?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5154

antonmosich commented:
Please squash your commits such that there's one commit adding you as a maintainer, and one commit adding connectorx.

coastalwhite commented:
> Please squash your commits such that there's one commit adding you as a maintainer, and one commit adding connectorx. Also, I'm encountering the same build error as @natsukium mentioned, so that should be fixed as well. Playing around locally, updating from 0.3.3 to 0.4.1 seems to change that error at least to the tests not working properly.

I updated it to 0.4.1, and now it is building properly again. I also squashed the commits.
